src/puppetlabs/puppetdb/amq_migration.clj

```diff
@@ -0,0 +1,369 @@
+
+(defn convert-old-command-format
+  "Converts from the command format that didn't include the certname
+  in the toplevel key list and had all the command data with the
+  payload key"
+  [message-bytes]
+  (let [{:keys [command version] :as old-command} (json/parse-strict message-bytes true)]
+    {:command command
+     :version version
+     :certname (or (get-in old-command [:payload :certname])
+                   (get-in old-command [:payload :name]))
+     :payload (coerce-clj->json-byte-stream (:payload old-command))}))
+
+(def upgrade-message-schema
+  {:command s/Str
+   :version s/Int
+   :certname s/Str
+   :payload java.io.ByteArrayInputStream})
+
+(s/defn coerce-to-new-command :-  upgrade-message-schema
+  "Commands coming out of this function are suitable for enqueuing
+  into stockpile. They must include command, version, certname and a
+  payload"
+  [{:keys [command version certname] :as headers} message-bytes]
+  (if (and command version certname)
+    {:command command
+     :version (Long/parseLong version)
+     :certname certname
+     :payload (java.io.ByteArrayInputStream. message-bytes)}
+    (convert-old-command-format message-bytes)))
+
+(defn discard-amq-message [message-bytes id headers exception dlo]
+  (let [now (ks/timestamp)]
+    (dlo/discard-bytes message-bytes
+                       id
+                       (or (:received headers)
+                           now)
+                       [{:exception exception
+                         :time now}]
+                       dlo)))
+
+(defn create-message-processor
+  [enqueue-fn dlo]
+  (fn [id ^Message message]
+    ;; When the queue is shutting down, it sends nil message
+    (when message
+      (let [headers (extract-headers message)
+            message-bytes (convert-message-body message)]
+        (try
+          (let [{:keys [command version certname payload] :as msg} (coerce-to-new-command headers message-bytes)]
+            (try
+              (enqueue-fn command version certname payload)
+              (catch Exception ex
+                (log/error ex (i18n/trs "[{0}] [{1}] [{2}] Unable to process message"
+                                        command version certname))
+                (discard-amq-message message-bytes id headers ex dlo))))
+          (catch AssertionError ex
+            (log/error ex (i18n/trs "Unable to process message: {0}" id))
+            (discard-amq-message message-bytes id headers ex dlo))
+          (catch Exception ex
+            (log/error ex (i18n/trs "Unable to process message: {0}" id))
+            (discard-amq-message message-bytes id headers ex dlo))
+          (finally
+            (.acknowledge message)))))))
+
+(defn consume-everything
+  "Attempts to read all available messages from ActiveMQ, giving up
+  after 5 seconds if one is not available"
+  [^MessageConsumer consumer process-message id]
+  (try
+    (loop []
+      (when-let [msg (.receive consumer 5000)]
```

Comment:
You're correct in your assessment here and it was intentional. Background here is the messages are already in the queue, we have properly started the broker, we are not allowing more messages to be enqueued. If there are no messages available, we must be done. Knowing we are done is the tricky part. The above code assumes that if we can't get a message in 5 seconds, we must be done. The time was totally arbitrary. From what I've read, the only way I have seen that is different, is to use the JMX metrics (which we know to be slightly out of date). It would be possible to combine the approaches. I'm open to that but I don't know that it would give me much more confidence in the correctness here. We can talk about some of the other things I've tried if you're interested @wkalt.